File: packages/smrt-android/library/src/main/kotlin/com/happyvertical/smrt/android/ui/MobileTheme.kt
package com.happyvertical.smrt.android.ui

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Shapes
import androidx.compose.material3.Text
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * SMRT mobile visual system, ported from amaru's `FieldOpsTheme` (ADR 0001
 * Phase 5) with the seed's intent intact: high-contrast, restrained,
 * durable — built for one-handed use in bright sun. Light surfaces,
 * near-black text, strong borders instead of soft shadows, and a small set
 * of literal status colors. Not a single-hue theme — the slate-blue primary
 * stays out of the way so the status colors carry meaning on their own.
 *
 * Productionized beyond the seed (which was light-only with global token
 * objects): a dark scheme exists but light stays the default (the field-use
 * posture), and status/spacing tokens are provided via CompositionLocals so
 * apps can override them per subtree (`MobileTheme.statusColors`,
 * `MobileTheme.spacing`).
 */

// Core palette (light).
private val Ink = Color(0xFF14181D) // primary text, near-black for contrast
private val InkMuted = Color(0xFF566069) // secondary text
private val Surface = Color(0xFFF5F6F8) // app background
private val CardSurface = Color(0xFFFFFFFF) // result / panel surface
private val CardMuted = Color(0xFFEDEFF2) // quiet fills, segmented controls
private val Line = Color(0xFFCFD5DC) // borders, dividers
private val Primary = Color(0xFF1F4E79) // serious slate-blue, used sparingly
private val OnPrimary = Color(0xFFFFFFFF)

// Core palette (dark) — same hue system, inverted value scale.
private val DarkInk = Color(0xFFE7EAEE)
private val DarkInkMuted = Color(0xFFA6AEB6)
private val DarkSurface = Color(0xFF14181D)
private val DarkCardSurface = Color(0xFF1C2127)
private val DarkCardMuted = Color(0xFF242A31)
private val DarkLine = Color(0xFF3A424B)
private val DarkPrimary = Color(0xFF9CC2EC)
private val DarkOnPrimary = Color(0xFF0F2A47)

/**
 * Literal status colors (alerts, translation state, permission state).
 * The five roles are the seed's `FieldOpsStatus` scale.
 */
@Immutable
data class MobileStatusColors(
    val stop: Color,
    val caution: Color,
    val go: Color,
    val info: Color,
    val neutral: Color,
)

val LightMobileStatusColors = MobileStatusColors(
    stop = Color(0xFFB3261E),
    caution = Color(0xFFA85A00),
    go = Color(0xFF1E7D34),
    info = Color(0xFF1F4E79),
    neutral = Color(0xFF566069),
)

val DarkMobileStatusColors = MobileStatusColors(
    stop = Color(0xFFE07A73),
    caution = Color(0xFFE0A050),
    go = Color(0xFF66BB6A),
    info = Color(0xFF9CC2EC),
    neutral = Color(0xFFA6AEB6),
)

/** Spacing scale. Generous touch targets and gaps for glove use. */
@Immutable
data class MobileSpacing(
    val screenPadding: androidx.compose.ui.unit.Dp = 20.dp,
    val sectionGap: androidx.compose.ui.unit.Dp = 18.dp,
    val itemGap: androidx.compose.ui.unit.Dp = 12.dp,
    val tightGap: androidx.compose.ui.unit.Dp = 6.dp,
    val minTouchTarget: androidx.compose.ui.unit.Dp = 56.dp,
)

val LocalMobileStatusColors = staticCompositionLocalOf { LightMobileStatusColors }
val LocalMobileSpacing = staticCompositionLocalOf { MobileSpacing() }

/** Token accessors, [MaterialTheme]-style: `MobileTheme.statusColors.stop`. */
object MobileTheme {
    val statusColors: MobileStatusColors
        @Composable get() = LocalMobileStatusColors.current

    val spacing: MobileSpacing
        @Composable get() = LocalMobileSpacing.current
}

private val MobileLightColorScheme = lightColorScheme(
    primary = Primary,
    onPrimary = OnPrimary,
    primaryContainer = Color(0xFFDCE6F1),
    onPrimaryContainer = Primary,
    secondary = Primary,
    onSecondary = OnPrimary,
    background = Surface,
    onBackground = Ink,
    surface = CardSurface,
    onSurface = Ink,
    surfaceVariant = CardMuted,
    onSurfaceVariant = InkMuted,
    outline = Line,
    outlineVariant = Line,
    error = LightMobileStatusColors.stop,
    onError = Color(0xFFFFFFFF),
)

private val MobileDarkColorScheme = darkColorScheme(
    primary = DarkPrimary,
    onPrimary = DarkOnPrimary,
    primaryContainer = Primary,
    onPrimaryContainer = Color(0xFFDCE6F1),
    secondary = DarkPrimary,
    onSecondary = DarkOnPrimary,
    background = DarkSurface,
    onBackground = DarkInk,
    surface = DarkCardSurface,
    onSurface = DarkInk,
    surfaceVariant = DarkCardMuted,
    onSurfaceVariant = DarkInkMuted,
    outline = DarkLine,
    outlineVariant = DarkLine,
    error = DarkMobileStatusColors.stop,
    onError = Color(0xFF3A0A07),
)

private val MobileTypography = Typography(
    headlineSmall = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 22.sp, lineHeight = 28.sp),
    titleLarge = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 19.sp, lineHeight = 24.sp),
    titleMedium = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 17.sp, lineHeight = 22.sp),
    bodyLarge = TextStyle(fontWeight = FontWeight.Normal, fontSize = 16.sp, lineHeight = 22.sp),
    bodyMedium = TextStyle(fontWeight = FontWeight.Normal, fontSize = 15.sp, lineHeight = 20.sp),
    bodySmall = TextStyle(fontWeight = FontWeight.Normal, fontSize = 13.sp, lineHeight = 18.sp),
    labelLarge = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 15.sp, lineHeight = 18.sp),
    labelMedium = TextStyle(fontWeight = FontWeight.SemiBold, fontSize = 12.sp, lineHeight = 16.sp),
)

private val MobileShapes = Shapes(
    extraSmall = RoundedCornerShape(6.dp),
    small = RoundedCornerShape(8.dp),
    medium = RoundedCornerShape(10.dp),
    large = RoundedCornerShape(12.dp),
)

/**
 * The SMRT mobile theme. [darkTheme] defaults to false deliberately — the
 * seed system is tuned for bright-sun outdoor legibility; apps that want to
 * follow the system pass `isSystemInDarkTheme()`.
 */
@Composable
fun MobileTheme(
    darkTheme: Boolean = false,
    statusColors: MobileStatusColors = if (darkTheme) DarkMobileStatusColors else LightMobileStatusColors,
    spacing: MobileSpacing = MobileSpacing(),
    content: @Composable () -> Unit,
) {
    CompositionLocalProvider(
        LocalMobileStatusColors provides statusColors,
        LocalMobileSpacing provides spacing,
    ) {
        MaterialTheme(
            colorScheme = if (darkTheme) MobileDarkColorScheme else MobileLightColorScheme,
            typography = MobileTypography,
            shapes = MobileShapes,
            content = content,
        )
    }
}

/**
 * Small section heading used to group scannable results. Quiet by design —
 * it labels a list, it is not an app title.
 */
@Composable
fun MobileSectionLabel(
    text: String,
    modifier: Modifier = Modifier,
    trailing: String? = null,
) {
    val label = if (trailing.isNullOrBlank()) text.uppercase() else "${text.uppercase()}  •  $trailing"
    Box(modifier = modifier) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

/**
 * Flat, bordered result container. Repeated result items are okay as cards;
 * shadows are not — borders read better in sun and keep the surface quiet.
 * Never nest one of these inside another.
 */
@Composable
fun MobileResultCard(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    Card(
        modifier = modifier,
        shape = MaterialTheme.shapes.medium,
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surface),
        elevation = CardDefaults.cardElevation(defaultElevation = 0.dp),
        border = BorderStroke(1.dp, MaterialTheme.colorScheme.outline),
        content = { content() },
    )
}

/** Compact status chip with a tinted fill and colored label. */
@Composable
fun MobileStatusPill(
    text: String,
    color: Color,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .background(color.copy(alpha = 0.12f), RoundedCornerShape(6.dp))
            .padding(horizontal = 8.dp, vertical = 3.dp),
    ) {
        Text(
            text = text,
            style = MaterialTheme.typography.labelMedium,
            color = color,
        )
    }
}

File: packages/smrt-ios/Sources/SmrtIos/Theme/MobileTheme.swift
import SwiftUI

// MARK: - Design tokens
//
// The SwiftUI half of the SMRT mobile design system. Token values are the
// SAME as smrt-android's `MobileTheme.kt` (ADR 0001 Phase 5) — porting from the
// already-productionized Android tokens keeps the two platforms token-identical
// rather than drifting. amaru's inline iOS `FieldOpsStyle` had only 4 colors and
// no stop/info/neutral; this is the productionized set.
//
// Delivery mirrors Compose's CompositionLocals with SwiftUI `@Environment`:
// `MobileColors`, `MobileStatusColors`, and `MobileSpacing` are environment
// values apps can override per-subtree, provided by the `MobileTheme` container.

// MARK: Core palette

/// Core surface/text palette. Light is the default (bright-sun, one-handed field
/// posture — the seed's deliberate choice); apps opt into `dark` via `MobileTheme`.
public struct MobileColors: Sendable, Equatable {
    public let ink: Color
    public let inkMuted: Color
    public let surface: Color
    public let cardSurface: Color
    public let cardMuted: Color
    public let line: Color
    public let primary: Color
    public let onPrimary: Color

    public init(
        ink: Color, inkMuted: Color, surface: Color, cardSurface: Color,
        cardMuted: Color, line: Color, primary: Color, onPrimary: Color
    ) {
        self.ink = ink
        self.inkMuted = inkMuted
        self.surface = surface
        self.cardSurface = cardSurface
        self.cardMuted = cardMuted
        self.line = line
        self.primary = primary
        self.onPrimary = onPrimary
    }

    public static let light = MobileColors(
        ink: Color(hex: 0x14181D),
        inkMuted: Color(hex: 0x566069),
        surface: Color(hex: 0xF5F6F8),
        cardSurface: Color(hex: 0xFFFFFF),
        cardMuted: Color(hex: 0xEDEFF2),
        line: Color(hex: 0xCFD5DC),
        primary: Color(hex: 0x1F4E79),
        onPrimary: Color(hex: 0xFFFFFF)
    )

    public static let dark = MobileColors(
        ink: Color(hex: 0xE7EAEE),
        inkMuted: Color(hex: 0xA6AEB6),
        surface: Color(hex: 0x14181D),
        cardSurface: Color(hex: 0x1C2127),
        cardMuted: Color(hex: 0x242A31),
        line: Color(hex: 0x3A424B),
        primary: Color(hex: 0x9CC2EC),
        onPrimary: Color(hex: 0x0F2A47)
    )
}

/// Literal status colors — Stop/Caution/Go/Info/Neutral — not decorative accents.
public struct MobileStatusColors: Sendable, Equatable {
    public let stop: Color
    public let caution: Color
    public let go: Color
    public let info: Color
    public let neutral: Color

    public init(stop: Color, caution: Color, go: Color, info: Color, neutral: Color) {
        self.stop = stop
        self.caution = caution
        self.go = go
        self.info = info
        self.neutral = neutral
    }

    public static let light = MobileStatusColors(
        stop: Color(hex: 0xB3261E),
        caution: Color(hex: 0xA85A00),
        go: Color(hex: 0x1E7D34),
        info: Color(hex: 0x1F4E79),
        neutral: Color(hex: 0x566069)
    )

    public static let dark = MobileStatusColors(
        stop: Color(hex: 0xE07A73),
        caution: Color(hex: 0xE0A050),
        go: Color(hex: 0x66BB6A),
        info: Color(hex: 0x9CC2EC),
        neutral: Color(hex: 0xA6AEB6)
    )
}

/// Glove-use spacing scale (points), matching Android's `MobileSpacing` (dp).
public struct MobileSpacing: Sendable, Equatable {
    public let screenPadding: CGFloat
    public let sectionGap: CGFloat
    public let itemGap: CGFloat
    public let tightGap: CGFloat
    public let minTouchTarget: CGFloat

    public init(
        screenPadding: CGFloat = 20,
        sectionGap: CGFloat = 18,
        itemGap: CGFloat = 12,
        tightGap: CGFloat = 6,
        minTouchTarget: CGFloat = 56
    ) {
        self.screenPadding = screenPadding
        self.sectionGap = sectionGap
        self.itemGap = itemGap
        self.tightGap = tightGap
        self.minTouchTarget = minTouchTarget
    }

    public static let `default` = MobileSpacing()
}

/// Type scale (size / weight), matching Android's `MobileTypography`.
public enum MobileTypography {
    public static let headlineSmall = Font.system(size: 22, weight: .semibold)
    public static let titleLarge = Font.system(size: 19, weight: .semibold)
    public static let titleMedium = Font.system(size: 17, weight: .semibold)
    public static let bodyLarge = Font.system(size: 16, weight: .regular)
    public static let bodyMedium = Font.system(size: 15, weight: .regular)
    public static let bodySmall = Font.system(size: 13, weight: .regular)
    public static let labelLarge = Font.system(size: 15, weight: .semibold)
    public static let labelMedium = Font.system(size: 12, weight: .semibold)
}

/// Corner radii, matching Android's `MobileShapes`.
public enum MobileShapes {
    public static let extraSmall: CGFloat = 6
    public static let small: CGFloat = 8
    public static let medium: CGFloat = 10
    public static let large: CGFloat = 12
}

// MARK: - Environment plumbing (the SwiftUI analog of Compose CompositionLocals)

private struct MobileColorsKey: EnvironmentKey {
    static let defaultValue = MobileColors.light
}
private struct MobileStatusColorsKey: EnvironmentKey {
    static let defaultValue = MobileStatusColors.light
}
private struct MobileSpacingKey: EnvironmentKey {
    static let defaultValue = MobileSpacing.default
}

public extension EnvironmentValues {
    var mobileColors: MobileColors {
        get { self[MobileColorsKey.self] }
        set { self[MobileColorsKey.self] = newValue }
    }
    var mobileStatusColors: MobileStatusColors {
        get { self[MobileStatusColorsKey.self] }
        set { self[MobileStatusColorsKey.self] = newValue }
    }
    var mobileSpacing: MobileSpacing {
        get { self[MobileSpacingKey.self] }
        set { self[MobileSpacingKey.self] = newValue }
    }
}

// MARK: - Theme container

/// Provides the design tokens to a subtree. `darkTheme` defaults to `false`
/// deliberately (bright-sun field posture); apps opt in — e.g.
/// `MobileTheme(darkTheme: colorScheme == .dark) { … }`.
public struct MobileTheme<Content: View>: View {
    private let colors: MobileColors
    private let statusColors: MobileStatusColors
    private let spacing: MobileSpacing
    private let content: Content

    public init(
        darkTheme: Bool = false,
        colors: MobileColors? = nil,
        statusColors: MobileStatusColors? = nil,
        spacing: MobileSpacing = .default,
        @ViewBuilder content: () -> Content
    ) {
        self.colors = colors ?? (darkTheme ? .dark : .light)
        self.statusColors = statusColors ?? (darkTheme ? .dark : .light)
        self.spacing = spacing
        self.content = content()
    }

    public var body: some View {
        content
            .environment(\.mobileColors, colors)
            .environment(\.mobileStatusColors, statusColors)
            .environment(\.mobileSpacing, spacing)
            .tint(colors.primary)
            .background(colors.surface.ignoresSafeArea())
    }
}

// MARK: - Widgets

/// Quiet uppercase group label, optional trailing note. Matches `MobileSectionLabel`.
public struct MobileSectionLabel: View {
    private let text: String
    private let trailing: String?
    @Environment(\.mobileColors) private var colors

    public init(_ text: String, trailing: String? = nil) {
        self.text = text
        self.trailing = trailing
    }

    public var body: some View {
        Text(labelText)
            .font(MobileTypography.labelMedium)
            .foregroundStyle(colors.inkMuted)
    }

    private var labelText: String {
        if let trailing, !trailing.isEmpty {
            return "\(text.uppercased())  •  \(trailing)"
        }
        return text.uppercased()
    }
}

/// Flat, bordered panel — borders read better in sun than shadows; never nested.
/// Matches `MobileResultCard`.
public struct MobileResultCard<Content: View>: View {
    private let content: Content
    @Environment(\.mobileColors) private var colors

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) { content }
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(colors.cardSurface)
            .clipShape(RoundedRectangle(cornerRadius: MobileShapes.medium))
            .overlay(
                RoundedRectangle(cornerRadius: MobileShapes.medium)
                    .stroke(colors.line, lineWidth: 1)
            )
    }
}

/// A 12%-alpha tinted fill with the full-strength color as the label. Matches
/// `MobileStatusPill`. Pass a role from the environment's `MobileStatusColors`.
public struct MobileStatusPill: View {
    private let text: String
    private let color: Color

    public init(_ text: String, color: Color) {
        self.text = text
        self.color = color
    }

    public var body: some View {
        Text(text)
            .font(MobileTypography.labelMedium)
            .foregroundStyle(color)
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            .background(color.opacity(0.12))
            .clipShape(RoundedRectangle(cornerRadius: MobileShapes.extraSmall))
    }
}

// MARK: - Color hex helper

extension Color {
    /// 0xRRGGBB → Color in sRGB, matching Android's `Color(0xFF……)` tokens.
    init(hex: UInt32) {
        self.init(
            .sRGB,
            red: Double((hex >> 16) & 0xFF) / 255.0,
            green: Double((hex >> 8) & 0xFF) / 255.0,
            blue: Double(hex & 0xFF) / 255.0,
            opacity: 1.0
        )
    }
}
